crates/swc_ecma_minifier/src/compress/optimize/switches.rs

```diff
@@ -16,171 +14,146 @@ where
+            for case in stmt.cases[(exact + 1)..].iter_mut() {
+                if terminate {
+                    var_ids.extend(case.cons.extract_var_ids())
+                } else {
+                    terminate |= case.cons.terminates();
+                    exact_case.cons.extend(case.cons.take())
                 }
             }
+            // remove default if there's an exact match
+            cases.retain(|case| case.test.is_some());
 
-            for case in &stmt.cases[..case_idx] {
-                for cons in &case.cons {
-                    var_ids.extend(
-                        cons.extract_var_ids()
-                            .into_iter()
-                            .map(|name| VarDeclarator {
-                                span: DUMMY_SP,
-                                name: Pat::Ident(name.into()),
-                                init: None,
-                                definite: Default::default(),
-                            }),
-                    );
+            if cases.len() == 2 {
+                let last = cases.last_mut().unwrap();
+
+                self.changed = true;
+                // so that following pass could turn it into if else
+                if let Some(test) = last.test.take() {
+                    prepend(&mut last.cons, test.into_stmt())
                 }
             }
+        }
 
-            for case in stmt.cases.iter_mut().skip(case_idx) {
-                let mut found_break = false;
-                case.cons.retain(|stmt| match stmt {
-                    Stmt::Break(BreakStmt { label: None, .. }) => {
-                        found_break = true;
-                        false
-                    }
+        if cases.len() == stmt.cases.len() {
+            stmt.cases = cases;
+            return;
+        }
 
-                    // TODO: Search recursively.
-                    Stmt::Break(BreakStmt {
-                        label: Some(break_label),
-                        ..
-                    }) => {
-                        if Some(break_label.to_id()) == label.as_ref().map(|label| label.to_id()) {
-                            found_break = true;
-                            false
-                        } else {
-                            !found_break
-                        }
-                    }
-                    _ => !found_break,
-                });
-
-                for case_stmt in case.cons.take() {
-                    match case_stmt {
-                        Stmt::Decl(Decl::Var(v)) if v.decls.iter().all(|v| v.init.is_none()) => {
-                            var_ids.extend(v.decls)
-                        }
-                        _ => {
-                            stmts.push(case_stmt);
-                        }
-                    }
-                }
-                if found_break {
-                    break;
-                }
-            }
+        self.optimize_switch_cases(&mut cases);
+
+        self.changed = true;
```

Comment:
Can you add a `report_change!`?

Reply:
Without it debugging infinite loop issues is so hard

Reply:
This is covered by L112 and L139